src/librustc_trans/collector.rs

```diff
@@ -456,12 +456,25 @@ impl<'a, 'tcx> MirVisitor<'tcx> for MirNeighborCollector<'a, 'tcx> {
         match *rvalue {
             mir::Rvalue::Aggregate(mir::AggregateKind::Closure(def_id,
                                                                ref substs), _) => {
-                assert!(can_have_local_instance(self.scx.tcx(), def_id));
-                let trans_item = create_fn_trans_item(self.scx.tcx(),
-                                                      def_id,
-                                                      substs.func_substs,
-                                                      self.param_substs);
-                self.output.push(trans_item);
+                let mir = errors::expect(self.scx.sess().diagnostic(), self.scx.get_mir(def_id),
+                                         || format!("Could not find MIR for closure: {:?}", def_id));
+
+                let concrete_substs = monomorphize::apply_param_substs(self.scx.tcx(),
+                                                                       self.param_substs,
+                                                                       &substs.func_substs);
+                let concrete_substs = self.scx.tcx().erase_regions(&concrete_substs);
+
+                let mut visitor = MirNeighborCollector {
```

Comment:
The difference is that we know do not create a separate `TransItem` for the closure and instead register neighbours for the enclosing function. This way we still find all monomorphizations and, as long as we make sure that closures are always instantiate together with their enclosing function, the inlinling map is also correct.
This will be changed back to having `TransItems` for closures when old-trans goes away.